File: typescript/src/taskManager.ts
/**
 * Gestionnaire de taches (version STARTER).
 *
 * ATTENTION ELEVES : ce module contient au moins UN BUG volontaire (deleteTask).
 * Vous le decouvrirez en Partie 4 (regression) du TP.
 */

import { Task } from "./task.js";
import { TaskNotFoundError } from "./errors.js";
import { validateStatus, type Priority, type Status } from "./validators.js";

const PRIORITY_WEIGHT: Record<Priority, number> = { low: 1, medium: 2, high: 3 };

export class TaskManager {
  private tasks: Task[] = [];
  private nextId = 1;

  createTask(params: {
    title: string;
    description?: string;
    priority?: Priority;
    dueDate?: string | null;
  }): Task {
    const task = new Task({
      id: this.nextId,
      title: params.title,
      description: params.description,
      priority: params.priority ?? "medium",
      status: "todo",
      dueDate: params.dueDate ?? null,
    });
    this.tasks.push(task);
    this.nextId += 1;
    return task;
  }

  listTasks(options?: {
    statusFilter?: Status;
    sortBy?: "priority" | "id" | "dueDate";
  }): Task[] {
    let result = [...this.tasks];
    if (options?.statusFilter !== undefined) {
      validateStatus(options.statusFilter);
      result = result.filter((t) => t.status === options.statusFilter);
    }
    const sortBy = options?.sortBy ?? "priority";
    if (sortBy === "priority") {
      result.sort((a, b) => PRIORITY_WEIGHT[b.priority] - PRIORITY_WEIGHT[a.priority]);
    } else if (sortBy === "id") {
      result.sort((a, b) => a.id - b.id);
    } else if (sortBy === "dueDate") {
      result.sort((a, b) => (a.dueDate ?? "9999").localeCompare(b.dueDate ?? "9999"));
    }
    return result;
  }

  getTask(taskId: number): Task {
    const task = this.tasks.find((t) => t.id === taskId);
    if (!task) throw new TaskNotFoundError(`Aucune tache trouvee avec id=${taskId}.`);
    return task;
  }

  markDone(taskId: number): Task {
    const task = this.getTask(taskId);
    task.status = "done";
    return task;
  }

  deleteTask(taskId: number): boolean {
    const index = this.tasks.findIndex((t) => t.id === taskId);
    if (index === -1) {
      throw new TaskNotFoundError(`Aucune tache trouvee avec id=${taskId}.`);
    }
    this.tasks.splice(index, 1);
    return true;
  }
  

  getStats(): { todo: number; doing: number; done: number; total: number } {
    const stats = { todo: 0, doing: 0, done: 0, total: this.tasks.length };
    for (const t of this.tasks) {
      stats[t.status] += 1;
    }
    return stats;
  }

  replaceAll(tasks: Task[]): void {
    this.tasks = [...tasks];
    this.nextId = tasks.length > 0 ? Math.max(...tasks.map((t) => t.id)) + 1 : 1;
  }
}

File: python/src/task_manager.py
"""Gestionnaire de taches : logique metier.

ATTENTION ELEVES : ce module contient au moins UN BUG volontaire.
Vous devez le decouvrir en Partie 4 (regression) en ecrivant un test qui echoue,
puis le corriger, puis garder le test pour blinder le code.
"""

from .task import Task
from .exceptions import TaskNotFoundError
from .validators import validate_status

PRIORITY_WEIGHT = {"low": 1, "medium": 2, "high": 3}


class TaskManager:
    def __init__(self):
        self._tasks = []
        self._next_id = 1

    def create_task(self, title, description="", priority="medium", due_date=None):
        task = Task(
            id=self._next_id,
            title=title,
            description=description,
            priority=priority,
            status="todo",
            due_date=due_date,
        )
        self._tasks.append(task)
        self._next_id += 1
        return task

    def list_tasks(self, status_filter=None, sort_by="priority"):
        result = list(self._tasks)
        if status_filter is not None:
            validate_status(status_filter)
            result = [t for t in result if t.status == status_filter]
        if sort_by == "priority":
            result.sort(key=lambda t: PRIORITY_WEIGHT[t.priority], reverse=True)
        elif sort_by == "id":
            result.sort(key=lambda t: t.id)
        elif sort_by == "due_date":
            result.sort(key=lambda t: (t.due_date is None, t.due_date or ""))
        return result

    def get_task(self, task_id):
        for task in self._tasks:
            if task.id == task_id:
                return task
        raise TaskNotFoundError(f"Aucune tache trouvee avec id={task_id}.")

    def update_task(self, task_id, **fields):
        task = self.get_task(task_id)
        for key, value in fields.items():
            if hasattr(task, key):
                setattr(task, key, value)
        # Revalidation complete pour garantir la coherence
        return Task(
            id=task.id,
            title=task.title,
            description=task.description,
            priority=task.priority,
            status=task.status,
            due_date=task.due_date,
            created_at=task.created_at,
        )

    def mark_done(self, task_id):
        task = self.get_task(task_id)
        task.status = "done"
        return task

    def delete_task(self, task_id):
        for i, task in enumerate(self._tasks):
            if task.id == task_id:
                self._tasks.pop(i)
                return True
        raise TaskNotFoundError(f"Aucune tache trouvee avec id={task_id}.")

    def get_stats(self):
        stats = {"todo": 0, "doing": 0, "done": 0, "total": len(self._tasks)}
        for task in self._tasks:
            stats[task.status] += 1
        return stats

    def replace_all(self, tasks):
        """Recharge l'integralite des taches depuis une liste (utile au chargement disque)."""
        self._tasks = list(tasks)
        self._next_id = max((t.id for t in self._tasks), default=0) + 1
